LibJS: Start implementing Date

linusg opened the pull request:
This adds:

- A global `Date` object (with `length` property and `now` function)
- The `Date` constructor (no arguments yet)
- The `Date` prototype (with `get*` functions)
- A bunch of tests

This is the most C++ I've ever written, so it's probably not perfect - I hope it's a good start however, and it works :^)

![image](https://user-images.githubusercontent.com/19366641/77864711-1d0bf980-7222-11ea-9fd3-65d9fb11993d.png)

xTibor commented:
```
Date: 2020-04-29
```

Time machine or an off-by-one error?

linusg commented:
> Time machine or an off-by-one error?

Ha, good catch - thanks! Off-by-one error, `Date.prototype.getMonth()` returns 0-11, `Core::DateTime` 1-12. Fixed :+1:

xTibor commented:
An another remark: For the `getXyz()` tests I'd rather use fixed timestamps instead of the current system time. Makes things more reproducible if you don't rely on external state.

awesomekling commented on Libraries/LibJS/Runtime/DatePrototype.cpp:
```diff
@@ -0,0 +1,145 @@
+    auto* this_object = interpreter.this_value().to_object(interpreter.heap());
+    if (!this_object)
+        return {};
+    ASSERT(this_object->is_date());
```
Let's make a helper function for turning `this` into a `DateObject`, something like this:

```cpp
DateObject* this_date_from_interpreter(Interpreter& interpreter)
{
    auto* this_object = interpreter.this_value().to_object(interpreter.heap());
    if (!this_object)
        return nullptr;
    if (!this_object->is_date()) {
        interpreter.throw_exception<Error>("TypeError", "this is not a DateObject");
        return nullptr;
    }
    return static_cast<DateObject*>(this_object);
}
```

WIth that, all the functions can call it up front:

```cpp
auto* this_object = this_date_from_interpreter(interpreter);
if (!this_object)
    return {};
...
```

awesomekling commented on Libraries/LibJS/Runtime/Date.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace JS {
+
+class Date : public Object {
```
Since this will not be subclasses, let's mark it `final`:
```suggestion
class Date final : public Object {
```

linusg commented:
@xTibor

> An another remark: For the `getXyz()` tests I'd rather use fixed timestamps instead of the current system time. Makes things more reproducible if you don't rely on external state.

Definitely, but as stated in the PR description neither `setXyz()` nor `Date` constructor parameters are implemented yet, so we have no way of using a fixed value (yet). :slightly_smiling_face:

@awesomekling thanks for the suggestions!

> JavaScript is specced to use double-precision floats internally, so we should stick to that for JS:Value.
IIRC doubles offer 53 bits of integer precision, so JavaScript can't natively represent 64-bit integers.
> In other words, let's not change JS::Value to use i64 as the implicit precision loss will likely lead to strange bugs down the line.

I'm not sure if there's a misunderstanding or if I'm just not getting it, but just to clarify: I didn't change `Value`'s internal `m_value.as_double` type to `i64`, that's still a `double` - I just added `explicit Value(i64 value)` and changed casts to `(i32)` to `(i64)`. Where is the implicit precision loss? :confused:

I was thinking that if you have a whole number that's exceeding the value of `i32` it'd wrap around - that's exactly what happened with (milliseconds) timestamps, as they are (now) larger than `2147483647`, and became negative numbers. If using `i64` isn't the right solution, what is?

awesomekling commented:
> > JavaScript is specced to use double-precision floats internally, so we should stick to that for JS:Value.
> > IIRC doubles offer 53 bits of integer precision, so JavaScript can't natively represent 64-bit integers.
> > In other words, let's not change JS::Value to use i64 as the implicit precision loss will likely lead to strange bugs down the line.
> 
> I'm not sure if there's a misunderstanding or if I'm just not getting it, but just to clarify: I didn't change `Value`'s internal `m_value.as_double` type to `i64`, that's still a `double` - I just added `explicit Value(i64 value)` and changed `(i32)` casts to `(i64)`. Where is the implicit precision loss? 
> 
> I was thinking that if you have a whole number that's exceeding the value of `i32` it'd wrap around - that's exactly what happened with (milliseconds) timestamps, as they are (now) larger than `2147483647`, and became negative numbers. If using `i64` isn't the right solution, what is?

The precision loss happens when we have an `i64` not representable as a `double` and store it into a `JS::Value`. The `i64` that comes out will be different than what we put in :)

The exact range of values we need to support in date math is specified here:
https://www.ecma-international.org/ecma-262/10.0/index.html#sec-time-values-and-time-range

My understanding is that it should be fine to simply use a `double` to represent these values.

awesomekling commented on Libraries/LibJS/Runtime/DatePrototype.cpp:
```diff
@@ -0,0 +1,149 @@
+
+namespace JS {
+
+Date* this_date_from_interpreter(Interpreter& interpreter)
```
Let's make this `static` since it's not used by anyone outside this .cpp file.

awesomekling commented on Libraries/LibJS/Runtime/DatePrototype.cpp:
```diff
@@ -0,0 +1,149 @@
+    auto* this_object = interpreter.this_value().to_object(interpreter.heap());
+    if (!this_object)
+        return nullptr;
+    if(!this_object->is_date()) {
```
Missing space after `if` here :)

linusg replied:
`clang-format`ted everything (again) :smile:

linusg commented:
> The precision loss happens when we have an i64 not representable as a double and store it into a JS::Value. The i64 that comes out will be different than what we put in :)

Thanks for the clarification! The value overflow I **saw** was coming from `Value::to_string()`, which casts the double value to `i32` before stringification :man_facepalming:
Sorry for the confusion. I removed the first comment.

linusg commented on Libraries/LibJS/Runtime/Value.cpp:
```diff
@@ -58,7 +58,7 @@ String Value::to_string() const
 
         // FIXME: This needs improvement.
-        return String::number((i32)as_double());
+        return String::format("%f", as_double());
```
Still needs improvement, just ensuring that numbers > `i32` are showing correctly.

linusg commented:
> Hmm, should we not be doing static_cast<double> rather than static_cast<i64> throughout?

Right - done!

awesomekling commented:
Whee! :)
